book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -0,0 +1,697 @@
+prevent database corruption, but it is the caller's responsibility (e.g. the
+zebra-state service's responsibility) to commit finalized blocks in order.
+
+2. Insert:
+    - `(hash, height)` into `height_by_hash`;
+    - `(height, hash)` into `hash_by_height`;
+    - `(height, block)` into `block_by_height`.
+
+3.  Update the `sprout_anchors` and `sapling_anchors` trees with the Sprout
+    and Sapling anchors (XXX: how??)
+
+4. Iterate over the enumerated transactions in the block. For each transaction:
+
+   1. Insert `(transaction_hash, block_height || BE32(tx_index))` to
+   `tx_by_hash`;
+
+   2. For each `TransparentInput::PrevOut { outpoint, .. }` in the
+   transaction's `inputs()`, remove `outpoint` from `utxo_by_output`.
+
+   3. For each `output` in the transaction's `outputs()`, construct the
+   `outpoint` that identifies it, and insert `(outpoint, output)` into
+   `utxo_by_output`.
+
+   4. For each [`JoinSplit`] description in the transaction,
+   insert `(nullifiers[0],())` and `(nullifiers[1],())` into
+   `sprout_nullifiers`.
+
+   5. For each [`Spend`] description in the transaction, insert
+   `(nullifier,())` into `sapling_nullifiers`.
+
+[`JoinSplit`]: https://doc.zebra.zfnd.org/zebra_chain/transaction/struct.JoinSplit.html
+[`Spend`]: https://doc.zebra.zfnd.org/zebra_chain/transaction/struct.Spend.html
+
+These updates can be performed in a batch or without necessarily iterating
+over all transactions, if the data is available by other means; they're
+specified this way for clarity.
+
+
+## Request / Response API
+[request-response]: #request-response
+
+The state API is provided by a pair of `Request`/`Response` enums. Each
+`Request` variant corresponds to particular `Response` variants, and it's
+fine (and encouraged) for caller code to unwrap the expected variants with
+`unreachable!` on the unexpected variants. This is slightly inconvenient but
+it means that we have a unified state interface with unified backpressure.
+
+This API includes both write and read calls. Spotting `Commit` requests in
+code review should not be a problem, but in the future, if we need to
+restrict access to write calls, we could implement a wrapper service that
+rejects these, and export "read" and "write" frontends to the same inner service.
+
+### `Request::CommitBlock`
+[request-commit-block]: #request-commit-block
+
+```rust
+CommitBlock {
+    block: Arc<Block>,
+    sprout_anchor: sprout::tree::Root,
+    sapling_anchor: sapling::tree::Root,
+}
+```
+
+Performs contextual validation of the given block, committing it to the state
+if successful. Returns `Response::Added(BlockHeaderHash)` with the hash of
+the newly committed block or an error.
+
+### `Request::CommitFinalizedBlock`
+[request-commit-finalized-block]: #request-finalized-block
+
+```rust
+CommitFinalizedBlock {
+    block: Arc<Block>,
+    sprout_anchor: sprout::tree::Root,
+    sapling_anchor: sapling::tree::Root,
```

Comment:
TODO: genesis anchors - use `None`?
https://github.com/ZcashFoundation/zebra/pull/902/files#r490675053

Reply:
It's not possible to spend any coins until 100 blocks after the genesis block, so by definition, the treestates are empty. We don't need to use None for them.
https://github.com/ZcashFoundation/zebra/pull/902/files#r494000878